// axios-chaos-interceptor/dist/createChaosInterceptor.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createChaosInterceptor = void 0;
var enhanceError_1 = require("./enhanceError");
var errors_1 = require("./errors");
var random = function (min, max) { return Math.random() * (max - min) + min; };
var percent = function (percent) { return random(0, 100) <= percent; };
var sleep = function (msec) {
    return new Promise(function (resolve) { return setTimeout(resolve, msec); });
};
var DEFAULT_ERROR_RATE = 90;
var DEFAULT_ERROR_MESSAGE = "Error caused by axios-chaos-interceptor";
var createChaosInterceptor = function (errors) {
    if (errors === void 0) { errors = errors_1.defaultErrors; }
    return function (res) { return __awaiter(void 0, void 0, void 0, function () {
        var _i, errors_2, error;
        var _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    _i = 0, errors_2 = errors;
                    _d.label = 1;
                case 1:
                    if (!(_i < errors_2.length)) return [3 /*break*/, 4];
                    error = errors_2[_i];
                    if (!percent((_a = error.rate) !== null && _a !== void 0 ? _a : DEFAULT_ERROR_RATE)) return [3 /*break*/, 3];
                    return [4 /*yield*/, sleep((_b = error.delay) !== null && _b !== void 0 ? _b : random(10, 1000))];
                case 2:
                    _d.sent();
                    throw enhanceError_1.createError({
                        message: "Request failed with status code " + error.status,
                        config: res.config,
                        code: null,
                        request: res.request,
                        response: __assign(__assign({}, res), { data: (_c = error.body) !== null && _c !== void 0 ? _c : DEFAULT_ERROR_MESSAGE, status: error.status }),
                    });
                case 3:
                    _i++;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/, res];
            }
        });
    }); };
};
exports.createChaosInterceptor = createChaosInterceptor;
