// axios-chaos-interceptor/dist/enhanceError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createError = exports.enhanceError = void 0;
var enhanceError = function (error, config, code, request, response) {
    error.config = config;
    if (code) {
        error.code = code;
    }
    error.request = request;
    error.response = response;
    error.isAxiosError = true;
    error.toJSON = function toJSON() {
        return {
            // Standard
            message: this.message,
            name: this.name,
            // Microsoft
            description: this.description,
            number: this.number,
            // Mozilla
            fileName: this.fileName,
            lineNumber: this.lineNumber,
            columnNumber: this.columnNumber,
            stack: this.stack,
            // Axios
            config: this.config,
            code: this.code,
        };
    };
    return error;
};
exports.enhanceError = enhanceError;
var createError = function (param) {
    var message = param.message, config = param.config, code = param.code, request = param.request, response = param.response;
    var error = new Error(message);
    return exports.enhanceError(error, config, code, request, response);
};
exports.createError = createError;
