// moleculer-ts/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const prettier_1 = __importDefault(require("prettier"));
const fs_1 = __importDefault(require("fs"));
const glob_1 = __importDefault(require("glob"));
const path_1 = __importDefault(require("path"));
const child_process_1 = __importDefault(require("child_process"));
const mustache_1 = __importDefault(require("mustache"));
const capitalize = (s) => {
    if (typeof s !== 'string') {
        return '';
    }
    return s.charAt(0).toUpperCase() + s.slice(1);
};
const servicesTemplate = fs_1.default.readFileSync(path_1.default.join(__dirname, 'templates', 'services.ts.mustache'), 'utf-8');
const brokerTemplate = fs_1.default.readFileSync(path_1.default.join(__dirname, 'templates', 'broker.ts.mustache'), 'utf-8');
const actionsParamsAssertTemplate = fs_1.default.readFileSync(path_1.default.join(__dirname, 'templates', 'actions.params.assert.ts.mustache'), 'utf-8');
const eventsParamsAssertTemplate = fs_1.default.readFileSync(path_1.default.join(__dirname, 'templates', 'events.params.assert.ts.mustache'), 'utf-8');
const rawMetaTemplate = fs_1.default.readFileSync(path_1.default.join(__dirname, 'templates/meta', 'raw.ts.mustache'), 'utf-8');
const namesMetaTemplate = fs_1.default.readFileSync(path_1.default.join(__dirname, 'templates/meta', 'names.ts.mustache'), 'utf-8');
async function formatAndSave(input, destination) {
    const info = await prettier_1.default.getFileInfo(destination);
    const options = (await prettier_1.default.resolveConfig(destination)) || undefined;
    if (options) {
        options.parser = info.inferredParser;
    }
    const output = prettier_1.default.format(input, options);
    await new Promise((resolve, reject) => {
        fs_1.default.mkdir(path_1.default.dirname(destination), { recursive: true }, err => {
            if (err) {
                reject(err);
                return;
            }
            resolve();
        });
    });
    await new Promise((resolve, reject) => {
        fs_1.default.writeFile(destination, output, err => {
            if (err) {
                reject(err);
                return;
            }
            resolve();
        });
    });
}
function getServiceTypeName(name) {
    const pureName = name
        .replace(/[^a-zA-Z0-9][a-zA-Z0-9]/g, one => one.charAt(0) + one.charAt(1).toUpperCase())
        .replace(/[^a-zA-Z0-9]/g, '');
    return `${capitalize(pureName)}ServiceTypes`;
}
function getRelativePathForImport(from, to) {
    return path_1.default.posix
        .relative(path_1.default.posix.normalize(from), path_1.default.posix.normalize(to))
        .replace(/\.ts$/, '');
}
async function rawMetaNames(services, outputDirImport) {
    // service types meta file content
    const metaFileContent = mustache_1.default.render(rawMetaTemplate, {
        serviceNames: services.map(({ name }) => getServiceTypeName(name)),
        outputDirImport,
    });
    const cpMeta = child_process_1.default.spawn(`${path_1.default.join('node_modules', '.bin', 'ts-node')}`, [
        '-e',
        metaFileContent,
    ]);
    let rawMeta = '';
    cpMeta.stdout.on('data', data => {
        rawMeta += data;
    });
    cpMeta.stderr.on('data', data => {
        console.error(`stderr: ${data}`);
    });
    await new Promise(resolve => {
        cpMeta.on('close', code => {
            resolve();
        });
    });
    const meta = JSON.parse(rawMeta);
    // broker action names
    const names = [];
    services.forEach(svc => {
        const { actionsLength, eventsLength } = meta[getServiceTypeName(svc.name)];
        names.push({
            name: getServiceTypeName(svc.name),
            actions: Array.from(Array(actionsLength).keys()),
            events: Array.from(Array(eventsLength).keys()),
        });
    });
    const metaNamesFileContent = mustache_1.default.render(namesMetaTemplate, {
        outputDirImport,
        names,
    });
    const cpMetaNames = child_process_1.default.spawn(`${path_1.default.join('node_modules', '.bin', 'ts-node')}`, ['-e', metaNamesFileContent]);
    let rawMetaNames = '';
    cpMetaNames.stdout.on('data', data => {
        rawMetaNames += data;
    });
    cpMetaNames.stderr.on('data', data => {
        console.error(`stderr: ${data}`);
    });
    await new Promise(resolve => {
        cpMetaNames.on('close', code => {
            resolve();
        });
    });
    return {
        meta,
        rawMetaNames: JSON.parse(rawMetaNames),
    };
}
async function generateBroker(options) {
    const isServiceName = options.isServiceName ||
        function (name) {
            return !Boolean(name.match(/^\$/));
        };
    const outputDirFs = path_1.default.normalize(options.outputDir);
    const outputDirImport = path_1.default.posix.normalize(options.outputDir);
    const serviceTypeFiles = glob_1.default.sync(options.serviceTypesPattern);
    const services = [];
    // init
    serviceTypeFiles.forEach(file => {
        const serviceRelativePath = getRelativePathForImport(options.outputDir, file);
        const service = require(file);
        const name = service.name;
        services.push({
            name,
            path: serviceRelativePath,
        });
    });
    // service types file content
    const serviceTypesFileContent = mustache_1.default.render(servicesTemplate, {
        services: services.map(({ path, name }) => {
            return {
                path,
                name: getServiceTypeName(name),
            };
        }),
    });
    await formatAndSave(serviceTypesFileContent, path_1.default.join(options.outputDir, 'services.types.ts'));
    const { meta, rawMetaNames: metaNames } = await rawMetaNames(services, outputDirImport);
    const callObj = {};
    const emitObj = {};
    // call
    services.forEach(svc => {
        const { actionsLength, eventsLength } = meta[getServiceTypeName(svc.name)];
        // actions GetCallParams/GetCallReturn
        for (let index = 0; index < actionsLength; index++) {
            const actionName = `${metaNames[`Services${getServiceTypeName(svc.name)}ActionsName${index}`]}`;
            const name = `${svc.name}.${actionName}`;
            if (callObj[name] !== undefined) {
                throw new Error(`Action ${name} multiple type definition detected.`);
            }
            callObj[name] = {
                actionName,
                name,
                index,
                type: getServiceTypeName(svc.name),
            };
        }
        // events GetEmitParams
        for (let index = 0; index < eventsLength; index++) {
            const eventName = `${metaNames[`Services${getServiceTypeName(svc.name)}EventsName${index}`]}`;
            const name = `${svc.name}.${eventName}`;
            if (emitObj[name] !== undefined) {
                throw new Error(`Event ${name} multiple type definition detected.`);
            }
            emitObj[name] = {
                eventName,
                name,
                index,
                type: getServiceTypeName(svc.name),
            };
        }
    });
    const brokerTypesFileContent = mustache_1.default.render(brokerTemplate, {
        callObj: Object.values(callObj),
        emitObj: Object.values(emitObj),
        ServiceNames: services
            .filter(({ name }) => isServiceName(name))
            .map(({ name }) => name),
        ServiceActionNames: Object.keys(callObj),
        ServiceEventNames: Object.keys(emitObj),
    });
    await formatAndSave(brokerTypesFileContent, path_1.default.join(outputDirFs, 'broker.types.ts'));
    if (options.generateActionsParamsAssert) {
        const servicesParamsAssertFileContent = mustache_1.default.render(actionsParamsAssertTemplate, {
            callObj: Object.values(callObj),
        });
        await formatAndSave(servicesParamsAssertFileContent, path_1.default.join(outputDirFs, 'actions.params.assert.ts'));
    }
    if (options.generateEventsParamsAssert) {
        const eventsParamsAssertFileContent = mustache_1.default.render(eventsParamsAssertTemplate, {
            emitObj: Object.values(emitObj),
        });
        await formatAndSave(eventsParamsAssertFileContent, path_1.default.join(outputDirFs, 'events.params.assert.ts'));
    }
}
exports.generateBroker = generateBroker;
